src/shims/unix/linux/sync.rs

```diff
@@ -85,16 +85,17 @@ pub fn futex<'tcx>(
                 return Ok(());
             }
 
-            // `deref_operand` but not actually dereferencing the ptr yet (it might be NULL!).
-            let timeout = this.ref_to_mplace(&this.read_immediate(&args[3])?)?;
-            let timeout_time = if this.ptr_is_null(timeout.ptr)? {
+            let timeout_op = &args[3];
+            let timeout_time = if this.ptr_is_null(this.read_pointer(timeout_op)?)? {
                 None
             } else {
                 if op & futex_realtime != 0 {
                     this.check_no_isolation(
                         "`futex` syscall with `op=FUTEX_WAIT` and non-null timeout with `FUTEX_CLOCK_REALTIME`",
                     )?;
                 }
+                let timeout =
+                    this.deref_operand_as(timeout_op, this.libc_ty_layout("timespec")?)?;
```

Comment:
Here it seems like we'd want a `deref_pointer_as`, so that you don't need to use `timeout_op` twice.